src/liboslexec/constfold.cpp

```diff
@@ -778,19 +779,34 @@ DECLFOLDER(constfold_aref)
     if (A.is_constant() && Index.is_constant()) {
         TypeSpec elemtype = A.typespec().elementtype();
         ASSERT (equivalent(elemtype, R.typespec()));
-        int index = *(int *)Index.data();
-        if (index < 0 || index >= A.typespec().arraylength()) {
-            // We are indexing a const array out of range.  But this
-            // isn't necessarily a reportable error, because it may be a
-            // code path that will never be taken.  Punt -- don't
-            // optimize this op, leave it to the execute-time range
-            // check to catch, if indeed it is a problem.
-            return 0;
-        }
-        ASSERT (index < A.typespec().arraylength());
+        int length = A.typespec().arraylength();
+        int index = *(int *)Index.data(), old_index = index;
+        if (index < 0) index = 0;
```

Comment:
Instead of these three lines, maybe:

    index = clamp(index, 0, length-1);

???